File: api/src/main/java/org/apache/gravitino/semantic/Dataset.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.semantic;

import com.google.common.base.Preconditions;
import java.util.Arrays;
import java.util.Objects;
import javax.annotation.Nullable;
import org.apache.gravitino.NameIdentifier;
import org.apache.gravitino.annotation.Evolving;

/**
 * An immutable dataset in a semantic model. A dataset binds a semantic name and optional semantic
 * metadata to a governed Gravitino table or logical view identified by {@link NameIdentifier}.
 */
@Evolving
public final class Dataset {

  private final String name;
  private final NameIdentifier source;

  @Nullable private final String[] primaryKey;
  @Nullable private final String[][] uniqueKeys;
  @Nullable private final String description;
  @Nullable private final AIContext aiContext;
  @Nullable private final Field[] fields;
  @Nullable private final CustomExtension[] customExtensions;

  private Dataset(Builder builder) {
    this.name = builder.name;
    this.source = builder.source;
    this.primaryKey = copyOrNull(builder.primaryKey);
    this.uniqueKeys = copyUniqueKeys(builder.uniqueKeys);
    this.description = builder.description;
    this.aiContext = builder.aiContext;
    this.fields =
        builder.fields == null ? null : Arrays.copyOf(builder.fields, builder.fields.length);
    this.customExtensions =
        builder.customExtensions == null
            ? null
            : Arrays.copyOf(builder.customExtensions, builder.customExtensions.length);
  }

  /**
   * Creates a builder for an immutable {@link Dataset}.
   *
   * @return A new builder.
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * Returns the dataset name.
   *
   * @return The dataset name.
   */
  public String name() {
    return name;
  }

  /**
   * Returns the governed table or logical view that backs this dataset.
   *
   * @return The source identifier.
   */
  public NameIdentifier source() {
    return source;
  }

  /**
   * Returns a copy of the primary key columns.
   *
   * @return The primary key columns, or {@code null} if they are not set.
   */
  @Nullable
  public String[] primaryKey() {
    return copyOrNull(primaryKey);
  }

  /**
   * Returns a deep copy of the unique key definitions.
   *
   * @return The unique keys, or {@code null} if they are not set.
   */
  @Nullable
  public String[][] uniqueKeys() {
    return copyUniqueKeys(uniqueKeys);
  }

  /**
   * Returns the dataset description.
   *
   * @return The dataset description, or {@code null} if it is not set.
   */
  @Nullable
  public String description() {
    return description;
  }

  /**
   * Returns the AI context associated with the dataset.
   *
   * @return The AI context, or {@code null} if it is not set.
   */
  @Nullable
  public AIContext aiContext() {
    return aiContext;
  }

  /**
   * Returns a copy of the fields defined by the dataset.
   *
   * @return The fields, or {@code null} if they are not set.
   */
  @Nullable
  public Field[] fields() {
    return fields == null ? null : Arrays.copyOf(fields, fields.length);
  }

  /**
   * Returns a copy of the custom extensions associated with the dataset.
   *
   * @return The custom extensions, or {@code null} if they are not set.
   */
  @Nullable
  public CustomExtension[] customExtensions() {
    return customExtensions == null
        ? null
        : Arrays.copyOf(customExtensions, customExtensions.length);
  }

  /**
   * Compares this dataset with another object for value equality.
   *
   * @param other The object to compare with.
   * @return {@code true} if the objects are equal, otherwise {@code false}.
   */
  @Override
  public boolean equals(@Nullable Object other) {
    if (this == other) {
      return true;
    }
    if (!(other instanceof Dataset)) {
      return false;
    }
    Dataset that = (Dataset) other;
    return name.equals(that.name)
        && source.equals(that.source)
        && Arrays.equals(primaryKey, that.primaryKey)
        && Arrays.deepEquals(uniqueKeys, that.uniqueKeys)
        && Objects.equals(description, that.description)
        && Objects.equals(aiContext, that.aiContext)
        && Arrays.equals(fields, that.fields)
        && Arrays.equals(customExtensions, that.customExtensions);
  }

  /**
   * Returns the value-based hash code for this dataset.
   *
   * @return The hash code.
   */
  @Override
  public int hashCode() {
    return Objects.hash(
        name,
        source,
        Arrays.hashCode(primaryKey),
        Arrays.deepHashCode(uniqueKeys),
        description,
        aiContext,
        Arrays.hashCode(fields),
        Arrays.hashCode(customExtensions));
  }

  /**
   * Returns a string representation of this dataset.
   *
   * @return The string representation.
   */
  @Override
  public String toString() {
    return "Dataset{"
        + "name='"
        + name
        + '\''
        + ", source="
        + source
        + ", primaryKey="
        + Arrays.toString(primaryKey)
        + ", uniqueKeys="
        + Arrays.deepToString(uniqueKeys)
        + ", description='"
        + description
        + '\''
        + ", aiContext="
        + aiContext
        + ", fields="
        + Arrays.toString(fields)
        + ", customExtensions="
        + Arrays.toString(customExtensions)
        + '}';
  }

  /** A builder for immutable {@link Dataset} values. */
  public static final class Builder {

    private String name;
    private NameIdentifier source;

    @Nullable private String[] primaryKey;
    @Nullable private String[][] uniqueKeys;
    @Nullable private String description;
    @Nullable private AIContext aiContext;
    @Nullable private Field[] fields;
    @Nullable private CustomExtension[] customExtensions;

    private Builder() {}

    /**
     * Sets the dataset name.
     *
     * @param name The non-empty dataset name.
     * @return This builder.
     */
    public Builder withName(String name) {
      this.name = name;
      return this;
    }

    /**
     * Sets the governed table or logical view that backs this dataset.
     *
     * @param source The source identifier.
     * @return This builder.
     */
    public Builder withSource(NameIdentifier source) {
      this.source = source;
      return this;
    }

    /**
     * Sets the optional primary key columns.
     *
     * @param primaryKey The primary key columns, or {@code null} to leave them unset.
     * @return This builder.
     */
    public Builder withPrimaryKey(@Nullable String[] primaryKey) {
      this.primaryKey = primaryKey;
      return this;
    }

    /**
     * Sets the optional unique key definitions.
     *
     * @param uniqueKeys The unique keys, or {@code null} to leave them unset.
     * @return This builder.
     */
    public Builder withUniqueKeys(@Nullable String[][] uniqueKeys) {
      this.uniqueKeys = uniqueKeys;
      return this;
    }

    /**
     * Sets the optional dataset description.
     *
     * @param description The description, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withDescription(@Nullable String description) {
      this.description = description;
      return this;
    }

    /**
     * Sets the optional AI context.
     *
     * @param aiContext The AI context, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withAIContext(@Nullable AIContext aiContext) {
      this.aiContext = aiContext;
      return this;
    }

    /**
     * Sets the optional semantic fields.
     *
     * @param fields The fields, or {@code null} to leave them unset.
     * @return This builder.
     */
    public Builder withFields(@Nullable Field[] fields) {
      this.fields = fields;
      return this;
    }

    /**
     * Sets the optional custom extensions.
     *
     * @param customExtensions The custom extensions, or {@code null} to leave them unset.
     * @return This builder.
     */
    public Builder withCustomExtensions(@Nullable CustomExtension[] customExtensions) {
      this.customExtensions = customExtensions;
      return this;
    }

    /**
     * Builds an immutable {@link Dataset}.
     *
     * @return The new dataset.
     * @throws IllegalArgumentException If the name is null or empty, the source is null, or an
     *     optional array contains an invalid element.
     */
    public Dataset build() {
      Preconditions.checkArgument(
          name != null && !name.isEmpty(), "name must not be null or empty");
      Preconditions.checkArgument(source != null, "source must not be null");
      SemanticModelDefinition.validateNonEmptyStringElements("primaryKey", primaryKey);
      validateUniqueKeys(uniqueKeys);
      SemanticModelDefinition.validateNoNullElements("fields", fields);
      SemanticModelDefinition.validateNoNullElements("customExtensions", customExtensions);
      return new Dataset(this);
    }
  }

  @Nullable
  private static String[] copyOrNull(@Nullable String[] values) {
    return values == null ? null : Arrays.copyOf(values, values.length);
  }

  @Nullable
  private static String[][] copyUniqueKeys(@Nullable String[][] uniqueKeys) {
    if (uniqueKeys == null) {
      return null;
    }

    String[][] copy = new String[uniqueKeys.length][];
    for (int index = 0; index < uniqueKeys.length; index++) {
      copy[index] = Arrays.copyOf(uniqueKeys[index], uniqueKeys[index].length);
    }
    return copy;
  }

  private static void validateUniqueKeys(@Nullable String[][] uniqueKeys) {
    if (uniqueKeys == null) {
      return;
    }
    for (int index = 0; index < uniqueKeys.length; index++) {
      String[] uniqueKey = uniqueKeys[index];
      Preconditions.checkArgument(
          uniqueKey != null && uniqueKey.length > 0,
          "uniqueKeys[%s] must not be null or empty",
          index);
      SemanticModelDefinition.validateNonEmptyStringElements(
          "uniqueKeys[" + index + "]", uniqueKey);
    }
  }
}

File: clients/client-python/gravitino/api/semantic/dataset.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from typing import Optional

from gravitino.api.semantic.ai_context import AIContext
from gravitino.api.semantic.custom_extension import CustomExtension
from gravitino.api.semantic.field import Field
from gravitino.api.semantic.semantic_utils import (
    check_no_none_elements,
    check_non_empty_string_elements,
)
from gravitino.name_identifier import NameIdentifier
from gravitino.utils.precondition import Precondition


class Dataset:  # pylint: disable=too-many-instance-attributes
    """A source-backed dataset exposed by a Semantic Model.

    Dataset names are unique within a Semantic Model. The source is a three-part
    `NameIdentifier` that must resolve to a table or a logical view in the same
    metalake, inline query sources are not supported.
    """

    def __init__(
        self,
        name: str,
        source: NameIdentifier,
        primary_key: Optional[list[str]] = None,
        unique_keys: Optional[list[list[str]]] = None,
        description: Optional[str] = None,
        ai_context: Optional[AIContext] = None,
        fields: Optional[list[Field]] = None,
        custom_extensions: Optional[list[CustomExtension]] = None,
    ):
        Precondition.check_argument(
            name is not None and name != "", "name must not be null or empty"
        )
        Precondition.check_argument(source is not None, "source must not be null")
        check_non_empty_string_elements("primaryKey", primary_key)
        _check_unique_keys(unique_keys)
        check_no_none_elements("fields", fields)
        check_no_none_elements("customExtensions", custom_extensions)

        self._name = name
        self._source = source
        self._primary_key = None if primary_key is None else list(primary_key)
        self._unique_keys = _copy_unique_keys(unique_keys)
        self._description = description
        self._ai_context = ai_context
        self._fields = None if fields is None else list(fields)
        self._custom_extensions = (
            None if custom_extensions is None else list(custom_extensions)
        )

    def name(self) -> str:
        """Returns the dataset name."""
        return self._name

    def source(self) -> NameIdentifier:
        """Returns the identifier of the table or view backing the dataset."""
        return self._source

    def primary_key(self) -> Optional[list[str]]:
        """Returns the primary key columns, or `None` if they are not set."""
        return None if self._primary_key is None else list(self._primary_key)

    def unique_keys(self) -> Optional[list[list[str]]]:
        """Returns the unique key column groups, or `None` if they are not set."""
        return _copy_unique_keys(self._unique_keys)

    def description(self) -> Optional[str]:
        """Returns the dataset description, or `None` if it is not set."""
        return self._description

    def ai_context(self) -> Optional[AIContext]:
        """Returns the AI context, or `None` if it is not set."""
        return self._ai_context

    def fields(self) -> Optional[list[Field]]:
        """Returns the dataset fields, or `None` if they are not set."""
        return None if self._fields is None else list(self._fields)

    def custom_extensions(self) -> Optional[list[CustomExtension]]:
        """Returns the custom extensions, or `None` if they are not set."""
        return (
            None if self._custom_extensions is None else list(self._custom_extensions)
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Dataset):
            return False
        return (
            self._name == other.name()
            and self._source == other.source()
            and self._primary_key == other.primary_key()
            and self._unique_keys == other.unique_keys()
            and self._description == other.description()
            and self._ai_context == other.ai_context()
            and self._fields == other.fields()
            and self._custom_extensions == other.custom_extensions()
        )

    def __hash__(self) -> int:
        return hash(
            (
                self._name,
                self._source,
                None if self._primary_key is None else tuple(self._primary_key),
                (
                    None
                    if self._unique_keys is None
                    else tuple(tuple(key) for key in self._unique_keys)
                ),
                self._description,
                self._ai_context,
                None if self._fields is None else tuple(self._fields),
                (
                    None
                    if self._custom_extensions is None
                    else tuple(self._custom_extensions)
                ),
            )
        )

    def __repr__(self) -> str:
        return (
            f"Dataset(name={self._name!r}, source={self._source!r}, "
            f"primaryKey={self._primary_key!r}, uniqueKeys={self._unique_keys!r}, "
            f"description={self._description!r}, aiContext={self._ai_context!r}, "
            f"fields={self._fields!r}, customExtensions={self._custom_extensions!r})"
        )


def _check_unique_keys(unique_keys: Optional[list[list[str]]]) -> None:
    if unique_keys is None:
        return
    for index, unique_key in enumerate(unique_keys):
        Precondition.check_argument(
            unique_key is not None and len(unique_key) > 0,
            f"uniqueKeys[{index}] must not be null or empty",
        )
        check_non_empty_string_elements(f"uniqueKeys[{index}]", unique_key)


def _copy_unique_keys(
    unique_keys: Optional[list[list[str]]],
) -> Optional[list[list[str]]]:
    if unique_keys is None:
        return None
    return [list(unique_key) for unique_key in unique_keys]
